Userland/Libraries/LibCore/ResourceImplementation.cpp

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2023, Andrew Kaster <akaster@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibCore/DirIterator.h>
+#include <LibCore/ResourceImplementation.h>
+#include <LibCore/ResourceImplementationFile.h>
+#include <LibCore/System.h>
+
+namespace Core {
+
+static OwnPtr<ResourceImplementation> s_the;
+
+void ResourceImplementation::install(OwnPtr<ResourceImplementation> the)
+{
+    s_the = move(the);
+}
+
+ResourceImplementation& ResourceImplementation::the()
+{
+    if (!s_the) {
+        install(make<ResourceImplementationFile>("/res"_string));
+    }
+    return *s_the;
+}
+
+NonnullOwnPtr<Resource> ResourceImplementation::make_resource(String full_path, NonnullOwnPtr<Core::MappedFile> file)
+{
+    return adopt_own(*new Resource(move(full_path), Resource::Scheme::Resource, move(file)));
+}
+
+NonnullOwnPtr<Resource> ResourceImplementation::make_resource(String full_path, ByteBuffer buffer)
+{
+    return adopt_own(*new Resource(move(full_path), Resource::Scheme::Resource, move(buffer)));
+}
+
+NonnullOwnPtr<Resource> ResourceImplementation::make_directory_resource(String full_path)
+{
+    return adopt_own(*new Resource(move(full_path), Resource::Scheme::Resource, Resource::DirectoryTag {}));
+}
+
+ErrorOr<NonnullOwnPtr<Resource>> ResourceImplementation::load_from_uri(StringView uri)
+{
+    StringView const file_scheme = "file://"sv;
+    StringView const resource_scheme = "resource://"sv;
+
+    if (uri.starts_with(resource_scheme))
+        return load_from_resource_scheme_uri(uri);
+
+    if (uri.starts_with(file_scheme)) {
+        auto path = uri.substring_view(file_scheme.length());
```

Comment:
This code looks like it wants to use URL.

Reply:
Not really, it only wants to inspect the front of the "path" for a specific scheme. Any fragments, queries, etc are invalid in resource names.